### PR TITLE
Fix #53 - Disambiguate Stop Codes

### LINKED ISSUE #53: Disambiguate duplicate stop codes
**Summary:** 

If you currently search for a `stop_code` in a region where more than one stop with that `stop_code` exists (usually where a region contains more than one agency), you'll get the message:

"I expected to find exactly one stop for the number you gave, but instead found 2.  This is not yet supported."

**Steps to reproduce:** 

Set region to Tampa, set stop to `3105`.

**Proposed solution:**

We could have Alexa enumerate the found stops, along with an identifier that we could capture to identify the stop they are interested in.

For example, if we defined a custom slot with "a, b, c, d...", we could read back the stops like:

"I found X stops with the number you gave.  Say "a" to select University Area Transit Center, "b" to select Fletcher and 42nd St., ..."

We should be able to store the stops response as an object in the `Session` variable, and then when the user responds retrieve the response from the `Session` and find the index for the stop based on the identifier spoken by the user.

@philipmw Thoughts?  Other ideas?

### PR DESCRIPTION
**Pull request by vikram31291**
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Sign the [electronic OBA ICLA](https://docs.google.com/forms/d/12jV-ByyN186MuPotMvxJtNKtSaGGTnEHm8rXomM2bm4/viewform)

- [x] Apply the [OneBusAway style template](https://github.com/OneBusAway/onebusaway/wiki/Code-Style) to your code.

- [x] Run the unit tests with `mvn test` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

File: interaction model/schema.json [hunk 1]
```diff
@@ -31,6 +31,12 @@
       ]
     },
     {
+      "intent": "AMAZON.YesIntent"
```
Comment (barbeau) [inline interaction model/schema.json]: Nit - could you please move these to the top of the schema, where the other `AMAZON.*` intents are?  I've been trying to keep the simple Amazon intents all grouped at the top.

File: src/main/java/org/onebusaway/alexa/AuthedSpeechlet.java [hunk 2]
```diff
@@ -79,7 +79,7 @@ public SpeechletResponse onIntent(final IntentRequest request,
             return anonSpeechlet.onIntent(request, session);
         } else if (GET_CITY.equals(intent.getName())) {
             return getCity();
-        } else if (SET_STOP_NUMBER.equals(intent.getName())) {
+        } else if (SET_STOP_NUMBER.equals(intent.getName()) || YES.equals(intent.getName()) || NO.equals(intent.getName())) {
```
Comment (barbeau) [inline src/main/java/org/onebusaway/alexa/AuthedSpeechlet.java]: Could you please split out the YES and NO intents to their own `else if` block?

`} else if (YES.equals(intent.getName()) || NO.equals(intent.getName()) {`

For now these are associated with the `SET_STOP_NUMBER` intent, but they could be more general in the future, and you've added code elsewhere to generalize them.

File: src/main/java/org/onebusaway/alexa/SessionAttribute.java [hunk 3]
```diff
@@ -27,4 +27,11 @@
     public static final String OBA_BASE_URL = "obaBaseUrl";
     public static final String PREVIOUS_RESPONSE = "previousResponse";
     public static final String LAST_ACCESS_TIME = "lastAccessTime";
+    public static final String FOUND_STOPS = "foundStops";
+    public static final String ASK_STATE = "askState";
+
+    public static enum AskState {
```
Comment (barbeau) [inline src/main/java/org/onebusaway/alexa/SessionAttribute.java]: I believe you can remove `static` modifier here.

File: src/main/java/org/onebusaway/alexa/AnonSpeechlet.java [hunk 4]
```diff
@@ -172,6 +180,74 @@ public void onSessionEnded(SessionEndedRequest sessionEndedRequest, Session sess
 
     }
 
+    private SpeechletResponse handleYesIntent(Session session, AskState askState) throws SpeechletException {
+        if (askState == AskState.VERIFYSTOP) {
+            return handleVerifyStopResponse(session, true /*stopFound*/);
+        }
+
+        log.error("Recieved yes intent without a question.");
```
Comment (barbeau) [inline src/main/java/org/onebusaway/alexa/AnonSpeechlet.java]: Typo - should be `Received`.

File: src/main/java/org/onebusaway/alexa/AnonSpeechlet.java [hunk 5]
```diff
@@ -172,6 +180,74 @@ public void onSessionEnded(SessionEndedRequest sessionEndedRequest, Session sess
...
+        return askForCity(Optional.empty());
+    }
+
+    private SpeechletResponse handleNoInent(Session session, AskState askState) throws SpeechletException {
```
Comment (barbeau) [inline src/main/java/org/onebusaway/alexa/AnonSpeechlet.java]: Typo - should be `handleNoIntent`.

File: src/main/java/org/onebusaway/alexa/AnonSpeechlet.java [hunk 6]
```diff
@@ -160,6 +164,10 @@ public SpeechletResponse onIntent(IntentRequest request, Session session)
             return setStopNumber(
                     stopNumberStr,
                     session);
+        } else if (YES.equals(intent.getName())) {
+            return handleYesIntent(session, askState);
+        } else if (NO.equals(intent.getName())) {
+            return handleNoInent(session, askState);
```
Comment (barbeau) [inline src/main/java/org/onebusaway/alexa/AnonSpeechlet.java]: Typo - should be `handleNoIntent`.

File: src/main/java/org/onebusaway/alexa/AnonSpeechlet.java [hunk 7]
```diff
@@ -172,6 +180,74 @@ public void onSessionEnded(SessionEndedRequest sessionEndedRequest, Session sess
...
+        log.error("Recieved yes intent without a question.");
+        return askForCity(Optional.empty());
+    }
+
+    private SpeechletResponse handleNoInent(Session session, AskState askState) throws SpeechletException {
+        if (askState == AskState.VERIFYSTOP) {
+            return handleVerifyStopResponse(session, false /*stopFound*/);
+        }
+
+        log.error("Recieved no intent without a question.");
```
Comment (barbeau) [inline src/main/java/org/onebusaway/alexa/AnonSpeechlet.java]: Typo - should be `Received`.

**Comment by barbeau** [comment]
Thanks @vikram31291!  This looks pretty good.  A few smaller comments inline.

My Echo power supply got fried and I'm waiting for a replacement, so I'm attempting to review this PR with Alex Dev Console Testing tab and the Echo Sim browser testing tool (which doesn't always seem to work well).  With these tools, I'm having intermittent issues getting it to recognize Yes/No responses.  For example, using EchoSim, if I set my region to "Tampa" and stop to "3105", it says that it found more than one stop, and asks if my stop is X.  However, if I say "Yes" or "No" it sometimes says:

>I could not locate a OneBusAway region near No, the city you gave...

In the Alexa Dev Console Testing tab, if I use lower case "yes" or "no" it works, but if I use upper case "Yes" or "No" it gives me the above region response.

Have you seen the same problem?  Or does it seem to work consitently for you on an actual Echo?

I'm not sure if this is an artifact of the limited testing tools I have, or if it's an issue on the actual Echo as well.

Also, we had [previously talked](https://github.com/OneBusAway/onebusaway-alexa/issues/53#issuecomment-220627720) about adding the agency name on the end based on the routes serving that stop.  So, for example:

"We found %d stops associated with the stop number. Did you mean the %s stop served by %s?"

Reasoning is that most duplicate stop codes arise from one stop code per agency, and the user will likely know which agency they want, but may not recognize the stop names.

Would you still be able to add this?

**Comment by barbeau** [comment]
ICLA manually confirmed :+1:

**Comment by vikram31291** [comment]
@barbeau I fixed the little issues and pushed that out. 

I will add the agency name in a separate change in the coming weeks. 

As for the yes/no issue, I will try on my echo tomorrow; currently don't have access to it. I was using the dev tools to verify.

**Comment by barbeau** [comment]
@vikram31291 I tested this using the Roger app and Yes/No recognition seems to work fine.  Any objections to going ahead and merging this?  I know we had talked about adding agency name but I'd like to go ahead and get this merged now, and we can add agency name in new PR.

**Comment by vikram31291** [comment]
@barbeau Sounds good to me!

**Comment by barbeau** [comment]
Thanks @vikram3129!  I appreciate you pulling this together.  Sorry it took so long to review.

**Comment by barbeau** [comment]
I opened a new issue for adding agency name at https://github.com/OneBusAway/onebusaway-alexa/issues/85.

**Comment by greenwoodcm** [comment]
I have the same issue that barbeau had.  When the app reads me the different stops and I say "yes" I get

"I could not locate a OneBusAway region near No, the city you gave..."

**Comment by barbeau** [comment]
@greenwoodcm bummer,  I had hoped this was fixed. Does it always give the region response,  or does it sometimes work correctly for you and set the stop?  And are you seeing this on an Echo?

**Comment by greenwoodcm** [comment]
I have never gotten it to work, always the same.  I can reproduce on both an echo and dot.

**Comment by barbeau** [comment]
@greenwoodcm Just opened new issue here - https://github.com/OneBusAway/onebusaway-alexa/issues/87.

**Pull request merged**